modules/system/defaults/dock.nix

```diff
@@ -128,18 +128,75 @@ in {
     };
 
     system.defaults.dock.persistent-apps = mkOption {
-      type = types.nullOr (types.listOf (types.either types.path types.str));
+      type = let
+        taggedType = types.attrTag {
+              app = mkOption {
+                description = "An application to be added to the dock.";
+                default = null;
+                type = types.submodule {
+                  options.path = mkOption {
+                    description = "Path to the application.";
+                    type = types.str;
+                  };
+                };
+              };
+              spacer = mkOption {
+                description = "A spacer to be added to the dock. Can be small or regular size.";
+                default = {
+                  small = false;
+                };
+                type = types.submodule {
+                  options.small = mkOption {
+                    description = "Whether the spacer is small.";
+                    type = types.bool;
+                  };
+                };
+              };
+              folder = mkOption {
+                description = "A folder to be added to the dock.";
+                default = null;
+                type = types.submodule {
+                  options.path = mkOption {
+                    description = "Path to the folder.";
+                    type = types.str;
+                  };
+                };
```

Comment:
Don't hold me to this, but assuming an app/folder will _only_ ever need a path value, and it's unlikely this submodule will ever expand, I think we could ditch the submodule and just assign a path directly?


```nix
folder = mkOption {
  description = "A folder to be added to the dock.";
  type = types.str;
  # (No default)
};
```

```nix
example = [
  { app = "/Applications/Safari.app"; }
  { spacer.small = false; }
  { spacer = { }; }
  { folder = "/System/Applications/Utilities"; }
]
```

Reply:
Yeah, I like this more... I was just trying to fulfill the requested form above. https://github.com/LnL7/nix-darwin/pull/1187#discussion_r1865152117

Reply:
I think a submodule makes sense if we expect there could be other keys to be set in future, e.g. if it was likely you may set `path` _and_ `name` or `icon` or whatever.

I'm not familiar with the `tile-data` schema so I don't know how likely that is.

Reply:
TBH I'm not very familiar with it either and the documentation isn't great either https://developer.apple.com/documentation/devicemanagement/dock/staticitem/tile-data.